shared/subspace-data-retrieval/src/object_fetcher.rs

```diff
@@ -0,0 +1,699 @@
+    },
+}
+
+/// A type-erased error
+pub type BoxError = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+/// Something that can be used to get decoded pieces by index
+#[async_trait]
+pub trait ObjectPieceGetter: fmt::Debug {
+    /// Get piece by index.
+    ///
+    /// Returns `Ok(None)` for temporary errors: the piece is not found, but immediately retrying
+    /// this provider might return it.
+    /// Returns `Err(_)` for permanent errors: this provider can't provide the piece at this time,
+    /// and another provider should be attempted.
+    async fn get_piece(&self, piece_index: PieceIndex) -> Result<Option<Piece>, BoxError>;
+}
+
+#[async_trait]
+impl<PG> ObjectPieceGetter for Arc<PG>
+where
+    PG: ObjectPieceGetter + Send + Sync + ?Sized,
+{
+    async fn get_piece(&self, piece_index: PieceIndex) -> Result<Option<Piece>, BoxError> {
+        self.as_ref().get_piece(piece_index).await
+    }
+}
+
+// Convenience methods, mainly used in testing
+#[async_trait]
+impl ObjectPieceGetter for NewArchivedSegment {
+    async fn get_piece(&self, piece_index: PieceIndex) -> Result<Option<Piece>, BoxError> {
+        if piece_index.segment_index() == self.segment_header.segment_index() {
+            return Ok(Some(
+                self.pieces
+                    .pieces()
+                    .nth(piece_index.position() as usize)
+                    .expect("checked segment index in if; piece must be present; qed"),
+            ));
+        }
+
+        Err(PieceGetterError::NotFound { piece_index }.into())
+    }
+}
+
+#[async_trait]
+impl ObjectPieceGetter for (PieceIndex, Piece) {
+    async fn get_piece(&self, piece_index: PieceIndex) -> Result<Option<Piece>, BoxError> {
+        if self.0 == piece_index {
+            return Ok(Some(self.1.clone()));
+        }
+
+        Err(PieceGetterError::NotFound { piece_index }.into())
+    }
+}
+
+// TODO: impl for IntoIterator instead?
+#[async_trait]
+impl ObjectPieceGetter for Vec<(PieceIndex, Piece)> {
+    async fn get_piece(&self, piece_index: PieceIndex) -> Result<Option<Piece>, BoxError> {
+        for (index, piece) in self.iter() {
+            if *index == piece_index {
+                return Ok(Some(piece.clone()));
+            }
+        }
+
+        Err(PieceGetterError::NotFound { piece_index }.into())
+    }
+}
+
+/// Object fetcher for the Subspace DSN.
+pub struct ObjectFetcher {
+    /// The piece getter used to fetch pieces.
+    piece_getter: Arc<dyn ObjectPieceGetter + Send + Sync + 'static>,
+
+    /// The maximum number of data bytes we'll read for a single object.
+    max_object_len: usize,
+}
+
+impl ObjectFetcher {
+    /// Create a new object fetcher with the given piece getter.
+    ///
+    /// `max_object_len` is the amount of data bytes we'll read for a single object before giving
+    /// up and returning an error, or `None` for no limit (`usize::MAX`).
+    pub fn new<PG>(piece_getter: PG, max_object_len: impl Into<Option<usize>>) -> Self
```

Comment:
Why `impl Into<Option<usize>>` and not just `Option<usize>`? Not sure what kind of input you expect that would satisfy such requirement and why.

Reply:
It accepts `usize` rather than `Some(usize)`, making the API more flexible and slightly more ergonomic.

But happy to switch to `Option<usize>`, see commit 51e4c7116b.